### DISCUSSION TITLE
How to use fastapi_users when foreign key constraints are not allowed?

### DISCUSSION THREAD

**Discussion by Sharjeel50** [fastapi-users/fastapi-users#1287] [Q&A] [answered]
Im using planet scale as my db as it has a lot of features I need, from what I've researched on the planetscale website (https://planetscale.com/docs/learn/operating-without-foreign-key-constraints), foreign key constraints are not allowed.

When attempting to create the oauth_account table I get the following error - 

```
sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (1105, 'VT10001: foreign key constraints are not allowed')
[SQL: 
CREATE TABLE oauth_account (
	user_id CHAR(36) NOT NULL, 
	id CHAR(36) NOT NULL, 
	oauth_name VARCHAR(100) NOT NULL, 
	access_token VARCHAR(1024) NOT NULL, 
	expires_at INTEGER, 
	refresh_token VARCHAR(1024), 
	account_id VARCHAR(320) NOT NULL, 
	account_email VARCHAR(320) NOT NULL, 
	PRIMARY KEY (id), 
	FOREIGN KEY(user_id) REFERENCES user (id)
)
```

Is there any changes I can make to fix this issue? would it be best to change the DB i am using so I can use foreign keys?

Here is the OAuthAccount below with user_id overridden, I've set ondelete to Set null as recommended in the planetscale mentioned above, but I still seen as Foreign key is still being used, instead of ForeignKey, can I change this to something else that would have the same effect? thanks

```
class OAuthAccount(SQLAlchemyBaseOAuthAccountTableUUID, Base):
    @declared_attr
    def user_id(cls) -> Mapped[GUID]:
        return mapped_column(
            GUID, ForeignKey("user.id", ondelete="SET NULL"), nullable=False
        )
```

**Comment by frankie567** [comment, marked as answer]
This is the first time I see a database provider promoting and boasting about *not supporting* foreign keys. IMHO, this is really... strange, to say the least but... well.

If you really want to use this service, you can inherit from our `SQLAlchemyBaseOAuthAccountTable` base model, which doesn't have the foreign key constraint. Then, you can do this:

* Define the `user_id` column as a plain UUID column.
* Define the SQLAlchemy ORM relationship manually by defining the joining method

```py
import uuid

from sqlalchemy.orm import relationship
from fastapi_users.db import SQLAlchemyBaseOAuthAccountTable
from fastapi_users_db_sqlalchemy.generics import GUID

class OAuthAccount(SQLAlchemyBaseOAuthAccountTable[UUID], Base):
    user_id: Mapped[UUID] = mapped_column(GUID)

    user = relationship(
        "User",
        backref="oauth_accounts",
        primaryjoin="OAuthAccount.user_id == User.id",
        foreign_keys=user_id,
    )
```

Ref: https://stackoverflow.com/questions/37806625/sqlalchemy-create-relations-but-without-foreign-key-constraint-in-db

**Reply by Sharjeel50** [reply]
Thanks for the quick response!

I tried this solution and was faced with the following error - `[Could not assemble any primary key columns for mapped table](https://stackoverflow.com/questions/24872541/could-not-assemble-any-primary-key-columns-for-mapped-table)`

but simply fixed this by adding `id: Mapped[UUID_ID] = mapped_column(GUID, primary_key=True, default=uuid.uuid4)` as a class variable, this is now working :)

heres the updated code if anyone faces the same issue - 

```
class OAuthAccount(SQLAlchemyBaseOAuthAccountTable[UUID], Base):
    id: Mapped[UUID_ID] = mapped_column(GUID, primary_key=True, default=uuid.uuid4)
    user_id: Mapped[UUID] = mapped_column(GUID)

    user = relationship(
        "User",
        backref="oauth_accounts",
        primaryjoin="OAuthAccount.user_id == User.id",
        foreign_keys=user_id,
    ),
```

Thanks for the help @frankie567 !
